File: Python/90.subsets-ii.py
#
# @lc app=leetcode id=90 lang=python
#
# [90] Subsets II
#
# https://leetcode.com/problems/subsets-ii/description/
#
# algorithms
# Medium (44.18%)
# Likes:    1150
# Dislikes: 56
# Total Accepted:    232.6K
# Total Submissions: 525.6K
# Testcase Example:  '[1,2,2]'
#
# Given a collection of integers that might contain duplicates, nums, return
# all possible subsets (the power set).
# 
# Note: The solution set must not contain duplicate subsets.
# 
# Example:
# 
# 
# Input: [1,2,2]
# Output:
# [
# ⁠ [2],
# ⁠ [1],
# ⁠ [1,2,2],
# ⁠ [2,2],
# ⁠ [1,2],
# ⁠ []
# ]
# 
# 
#

# @lc code=start
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """


        # # 迭代
        # result = [[]]
        # nums.sort()
        # for num in nums:
        #     temp = result.copy()
        #     for item in temp:
        #         if item + [num] not in result:
        #             result += [item+[num]]
        # return result


        # 递归
    #     nums.sort()
    #     result = []
    #     self.dfs(nums, 0, [], result)
    #     return result

    # def dfs(self, nums, index, path, result):
    #     if path not in result:
    #         result.append(path)
    #     for i in range(index, len(nums)):
    #         self.dfs(nums, i+1, path+[nums[i]], result)

        # 位操作
        nums.sort()
        result = []
        for i in range(1<<len(nums)):
            temp = []
            for j in range(len(nums)):
                if i & 1 << j:
                    temp.append(nums[j])
            if temp not in result:
                result.append(temp)
        return result

# @lc code=end

sol = Solution()
nums = [1,2,2]
print(sol.subsetsWithDup(nums))

File: C++/90.subsets-ii.cpp
/*
 * @lc app=leetcode id=90 lang=cpp
 *
 * [90] Subsets II
 *
 * https://leetcode.com/problems/subsets-ii/description/
 *
 * algorithms
 * Medium (44.51%)
 * Likes:    1235
 * Dislikes: 56
 * Total Accepted:    239.7K
 * Total Submissions: 536.7K
 * Testcase Example:  '[1,2,2]'
 *
 * Given a collection of integers that might contain duplicates, nums, return
 * all possible subsets (the power set).
 * 
 * Note: The solution set must not contain duplicate subsets.
 * 
 * Example:
 * 
 * 
 * Input: [1,2,2]
 * Output:
 * [
 * ⁠ [2],
 * ⁠ [1],
 * ⁠ [1,2,2],
 * ⁠ [2,2],
 * ⁠ [1,2],
 * ⁠ []
 * ]
 * 
 * 
 */
#include<vector>
using namespace std;

// @lc code=start
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {

        sort(nums.begin(), nums.end());

        vector<vector<int>> res;
        vector<int> path;
        DFS(nums, 0, path, res);
        return res;
    }

    void DFS(vector<int>& nums, int start, vector<int>& path, vector<vector<int>>& res) {
        // if (find(res.begin(), res.end(), path) == res.end())
        res.push_back(path);
        for (int i=start; i<nums.size(); i++) {
                if(i==start || nums[i]!=nums[i-1]){
                path.push_back(nums[i]);
                DFS(nums, i+1, path, res);
                path.pop_back();
            }
        }
    }
};
// @lc code=end

